[AArch64] Enable unrolling for small multi-exit loops

david-arm opened the pull request:
It can be highly beneficial to unroll small, two-block search loops
that look for a value in an array. An example of this would be
something that uses std::find to find a value in libc++. Older
versions of std::find in the libstdc++ headers are manually unrolled
in the source code, but this might change in newer releases where
the compiler is expected to either vectorise or unroll itself.

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4237,6 +4242,70 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  return true;
+}
+
+static bool shouldUnrollSmallMultiExitLoop(Loop *L, ScalarEvolution &SE,
```
This looks like a subset of `getAppleRuntimeUnrollPreferences`, can we unify it?

david-arm replied:
Sure, I could do that although it's actually a superset because in `getAppleRuntimeUnrollPreferences` it seems to disallow loops with multiple exit blocks, i.e. both exiting blocks must have the same destination.

david-arm replied:
Hmm, sorry, maybe that's not right. I'm just looking at this code:

```
  // Small search loops with multiple exits can be highly beneficial to unroll.
  if (!L->getExitBlock()) {
```

Perhaps I've misunderstood what this is doing?

david-arm replied:
I thought at first this is saying if the loop doesn't have a single exit block then unroll, however looking at the implementation of getExitBlock a bit more I can see that it doesn't permit repeats. So I guess that means it should always return false for any early exit loop.

fhahn replied:
The code block should handle small multi exit loops as advertised (tests are in `llvm/test/Transforms/LoopUnroll/AArch64/apple-unrolling-multi-exit.ll`). `getExitBlock()` should return the single exit block if the loop has a single exit & exiting block. 

If there are multiple exiting blocks to the same exit, the exit block will be included multiple times in getExitBlocks(); This is different from `getUniqueExitBlock`, which would return the common exit block, even if there are multiple exiting blocks branching to it.

david-arm replied:
Hi, I'm not sure exactly what you meant by 'unify' and how far I should attempt to commonise code. I've tried my best - see what you think! The main difference is that now getAppleRuntimeUnrollPreferences bails out on multi-exit loops and defers to common code. The common code bails out for slightly more cases, for example if an intrinsic gets lowered to a call, if the loop already contains a vector type, or if a cost cannot be calculated. I assume that these are sensible things to do for all targets anyway. One side-effect of these changes is that it's now possible to override the default unroll factor of 4 - I've added a test to Transforms/LoopUnroll/AArch64/apple-unrolling-multi-exit.ll for forcing unroll=2.

david-arm commented:
Hi @fhahn, I've tried to address your comments around trying to unify the code. Is it what you were expecting? Thanks!

igogo-x86 commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4237,6 +4242,81 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  if (I.getType()->isVectorTy())
+    return false;
+
+  if (isa<CallInst>(I) || isa<InvokeInst>(I)) {
```
Since we are only working with `CallBase` below, we can rewrite it as:
```
if (auto *CB = dyn_cast<CallBase>(&I)) {
  if (const Function *F = CB->getCalledFunction())
     ....
}
```

david-arm replied:
Hmm, it's a good suggestion but I checked and it seems `CallBrInst` also derives from `CallBase`. I'm not sure if I should implicitly add support for that by permitting all instances of `CallBase`. If you don't mind I'll leave this as it is for now?

david-arm replied:
Yep, you're right! I've fixed it now.

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4237,6 +4242,81 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  // Small search loops with multiple exits can be highly beneficial to unroll.
+  // We only care about loops with exactly two exiting blocks, although each
+  // block could jump to the same exit block.
+  SmallVector<BasicBlock *> Blocks(L->getBlocks());
```
Done

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -68,6 +68,11 @@ static cl::opt<bool> EnableOrLikeSelectOpt("enable-aarch64-or-like-select",
                                       cl::init(true), cl::Hidden);
 
+static cl::opt<unsigned> SmallMultiExitLoopUF(
+    "small-multi-exit-loop-unroll-factor", cl::init(0), cl::Hidden,
```
The name makes it sounds like a generally applicable flag, but it is something specific for the AArch64 cost model. I think ideally we would set things correctly per sub-target directly, which sounds like you are planning anyways, instead of adding this option first..

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4370,6 +4375,81 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  return true;
+}
+
+static InstructionCost getSizeOfLoop(Loop *L, AArch64TTIImpl &TTI) {
```
Can this also be shared with `getAppleRuntimeUnrollPreferences`? Although there we explicitly don't bail out on vector instructions. We only bail out on loops LV already vectorized.

david-arm replied:
Done!

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4370,6 +4375,81 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+    return false;
+
+  // Estimate the size of the loop.
+  InstructionCost Size = getSizeOfLoop(L, TTI);
+  if (!Size.isValid())
+    return false;
+
+  // Small search loops with multiple exits can be highly beneficial to unroll.
+  // We only care about loops with exactly two exiting blocks, although each
+  // block could jump to the same exit block.
+  ArrayRef<BasicBlock *> Blocks = L->getBlocks();
+  if (Blocks.size() != 2)
+    return false;
+
+  if (any_of(Blocks, [](BasicBlock *BB) {
+        return !isa<BranchInst>(BB->getTerminator());
+      }))
+    return false;
+
+  return *Size.getValue() < 6;
```
I guess that would make most sense to share this with `getAppleRuntimeUnrollPreferences`, as is at the moment the checks above are duplicated.

fhahn reviewed (commented):
I am not sure about adding a very specialized option for this. Is there a reason we couldnt take the right decision per uarch to start with? It sounds like that is planned down the road already?

david-arm commented:
> I am not sure about adding a very specialized option for this. Is there a reason we couldnt take the right decision per uarch to start with? It sounds like that is planned down the road already?

Well I was thinking that someone might be targeting a generic CPU with a particular architecture (perhaps this happens when targeting Android?), but still want a way to get the benefit of unrolling multi-exit loops such as std::find. Having said that, if the consensus is to avoid introducing a flag I can remove it?

davemgreen commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4370,6 +4375,81 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+
+      SmallVector<const Value *, 4> Operands(I.operand_values());
+      InstructionCost Cost =
+          TTI.getInstructionCost(&I, Operands, TTI::TCK_CodeSize);
```
This is done in some other places already (The unroller uses CodeMetrics to get the LoopSize). If we are not going to use that and our budget is quite small (6), then can we loop until we are over the budget then return false? It will help with larger loops.
Also just use `TTI.getInstructionCost(&I, TTI::TCK_CodeSize);` to let the function grab the operands.

david-arm replied:
With regards to the second comment ...

I tried using `TTI.getInstructionCost(&I, TTI::TCK_CodeSize)`, but annoyingly it's not available because the TTI object is of type AArch64TTIImpl, which seems to derive from  TargetTransformInfoImpl. Sadly TargetTransformInfoImpl only has the 3 operand version of getInstructionCost. :(

davemgreen reviewed (commented):
> I was thinking that someone might be targeting a generic CPU with a particular architecture (perhaps this happens when targeting Android?), but still want a way to get the benefit of unrolling multi-exit loops such as std::find. Having said that, if the consensus is to avoid introducing a flag I can remove it?

Having the option sounds useful for debugging - we can use it to disable the unrolling if we find issues. Perhaps rename it to -aarch64-small-multi-exit-loop-unroll-factor?

I gave this a try and it seemed OK. The code size, which was the thing I was worried about, did not increase by too much.

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4370,6 +4375,81 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+      // or for some operations that require SVE.
+      if (!Cost.isValid())
+        return InstructionCost::getInvalid();
+      Size += *Cost.getValue();
```
Done

davemgreen reviewed (commented):
This looks OK to me if @fhahn is OK with the new code.

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4375,6 +4380,87 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  return true;
+}
+
+static unsigned getLoopSize(Loop *L, AArch64TTIImpl &TTI,
```
Would also be good to document.

david-arm replied:
I've renamed this to `isLoopSizeWithinBudget` so that it returns a boolean instead, and passes the final size in an optional argument. I've also added some comments to the function. Hope this is better!

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4375,6 +4380,87 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+        return 0;
+
+      LoopCost += Cost;
+      if (LoopCost > Budget)
```
Could just decrement budget and check if it goes below 0?

david-arm replied:
Well, I may need to return the total size anyway which is why I wrote the code this way. If I decrement the budget and check against 0, then I still also need to increment the LoopCost so I figured perhaps better to leave as is?

fhahn reviewed (commented):
> > I am not sure about adding a very specialized option for this. Is there a reason we couldnt take the right decision per uarch to start with? It sounds like that is planned down the road already?
> 
> Well I was thinking that someone might be targeting a generic CPU with a particular architecture (perhaps this happens when targeting Android?), but still want a way to get the benefit of unrolling multi-exit loops such as std::find. Having said that, if the consensus is to avoid introducing a flag I can remove it?

I don't think we generally encourage users to use internal LLVM options, they mostly for LLVM developers, unless expose explicitly via `Clang` or other drivers. And those flags are also not guaranteed to be stable.

Another alternative as middle ground would be to introduce a target feature instead of a flag, this can be both specified via the command line and it will be easy to add it for CPUs where it's profitable.

But I think it would still be good to enable it for the CPUs that are known to benefit straigth away.

> Having the option sounds useful for debugging - we can use it to disable the unrolling if we find issues. Perhaps rename it to -aarch64-small-multi-exit-loop-unroll-factor?
> 
> I gave this a try and it seemed OK. The code size, which was the thing I was worried about, did not increase by too much.

If it is just about disabling unrolling for such loops, it might be good to check if the existing `unroll-runtime-multi-exit` option is explicitly set to false? This is something the current implementation for Apple CPUs doesn't do, but I think it would be reasonable to expect this flag to override the target specific decisions?

paulwalker-arm commented:
I don't believe we want or should be adding a new "public" way to control this optimisation and so we shouldn't add anything binding like a target feature flag.  The intent is really about having finer control for environments where it would not be easy to inject a new toolchain but a user is happy to try "secret" options using their existing toolchain, be it for functional or performance investigations.

I do agree we should be more aggressive in enabling this optimisation by default for the platforms we know it will benefit.

@fhahn: What is the down side to adding a dedicated option for this? If there's a problem with adding such flags then perhaps we should think about adding something akin to gcc's `-param` flag? Would there be benefit to having one flag with many properties verse having many flags?

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4389,43 +4475,15 @@ getAppleRuntimeUnrollPreferences(Loop *L, ScalarEvolution &SE,
       (SE.getSmallConstantMaxTripCount(L) > 0 &&
        SE.getSmallConstantMaxTripCount(L) <= 32))
     return;
-  if (findStringMetadataForLoop(L, "llvm.loop.isvectorized"))
```
Hmm, you're right I was too aggressive in removing it. I've re-added it after the check for early exits.

david-arm commented:
Following reviewer suggestions I've just gone ahead and enabled this for neoverse-v1 and neoverse-v2. The testing I've done doesn't suggest any downsides in terms of performance for SPEC2017 or any significant increase in code size. I saw less than 0.1% increase in code size for SPEC2017 and I observed a 0.3% increase for the LLVM test suite. However, when building SPEC2017 with libc++ I saw about 6% and 9% performance improvements for xalancbmk on neoverse-v1 and neoverse-v2, respectively.

davemgreen commented:
The option is just for debugging, it is not intended to be customer facing. Just in case we need to disable it again. If -unroll-runtime-multi-exit will reliably turn the option off then that would be be good. It doesn't look like that is how the option works though as it will not override ULO.RuntimeUnrollMultiExit.

> Following reviewer suggestions I've just gone ahead and enabled this for neoverse-v1 and neoverse-v2. The testing I've done doesn't suggest any downsides in terms of performance for SPEC2017 or any significant increase in code size. I saw less than 0.1% increase in code size for SPEC2017 and I observed a 0.3% increase for the LLVM test suite. However, when building SPEC2017 with libc++ I saw about 6% and 9% performance improvements for xalancbmk on neoverse-v1 and neoverse-v2, respectively.

We should enable for all cpus unless we have a good reason not to. We know this is relatively small impact on codesize, helps with big cores and little cores like unrolling (providing it is generally less instructions). Unless we have some reason not to, we should just enable it universally. If it does for some reason need to be per cpu then it should be a subtarget feature.

paulwalker-arm commented:
If the intention is to enable this for all AArch64 targets then yes I agree this means we only need to be able to disable the option. To avoid confusion I will clarify my "we don't want a target feature" comment.  Here I'm referring to public flags and specifically mean we don't want something at the -march level. `-target-feature` is effectively a secret flag and so I'm not against that.

fhahn commented:
> The option is just for debugging, it is not intended to be customer facing. Just in case we need to disable it again. If -unroll-runtime-multi-exit will reliably turn the option off then that would be be good. It doesn't look like that is how the option works though as it will not override ULO.RuntimeUnrollMultiExit.

I think this was something that was missed when adding `RuntimeUnrollMultiExit`. I think it makes sense for the option to override the profitability heuristic from TTI, same as it overrides the generic profitability heuristics: https://github.com/llvm/llvm-project/pull/134259

david-arm committed:
[AArch64] Enable unrolling for small multi-exit loops

It can be highly beneficial to unroll small, two-block search loops
that look for a value in an array. An example of this would be
something that uses std::find to find a value in libc++. Older
versions of std::find in the libstdc++ headers are manually unrolled
in the source code, but this might change in newer releases where
the compiler is expected to either vectorise or unroll itself.

david-arm commented:
I've now enabled unrolling of small, multi-exit loops by default for all cores and removed the extra flag I added previously, in favour of the existing `-unroll-runtime-multi-exit`.

davemgreen reviewed (approved):
LGTM

david-arm committed:
Address review comment

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4528,6 +4528,95 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  InstructionCost LoopCost = 0;
+
+  if (findStringMetadataForLoop(L, "llvm.loop.isvectorized"))
+    return 0;
```
Yep, good spot. Fixed now, thanks!

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4703,22 +4769,24 @@ void AArch64TTIImpl::getUnrollingPreferences(Loop *L, ScalarEvolution &SE,
     break;
   }
 
+  if (!L->getExitBlock() && shouldUnrollMultiExitLoop(L, SE, *this)) {
```
Could you add a comment explaining what loops we try to target here?

david-arm replied:
Yep, I've tried to add something.

fhahn commented on llvm/test/Transforms/LoopUnroll/AArch64/apple-unrolling-multi-exit.ll:
```diff
@@ -3,7 +3,7 @@
 ; RUN: opt -p loop-unroll -mcpu=apple-m4 -S %s | FileCheck --check-prefix=APPLE %s
-; RUN: opt -p loop-unroll -mcpu=apple-m1 -unroll-runtime-multi-exit=false -S %s | FileCheck --check-prefix=OTHER %s
+; RUN: opt -p loop-unroll -mcpu=apple-m1 -unroll-runtime-multi-exit=false -S %s | FileCheck --check-prefix=NOUNROLL %s
 ; RUN: opt -p loop-unroll -mcpu=cortex-a57 -S %s | FileCheck --check-prefix=OTHER %s
```
I think that line can be dropped now, as other CPU behavior is now covered by other tests

david-arm replied:
Assuming I've understood which line you meant, I've deleted the `-mcpu=cortex-a57` line.

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4528,6 +4528,95 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  // Estimate the size of the loop.
+  InstructionCost LoopCost = 0;
+
+  if (findStringMetadataForLoop(L, "llvm.loop.isvectorized"))
+    return false;
+
+  for (auto *BB : L->getBlocks()) {
+    for (auto &I : *BB) {
+      if (!shouldUnrollLoopWithInstruction(I, TTI))
```
Would it make more sense to check this separately? If not,  would be good to update the name to make it clearer that this does not only check if the loop size is within a budget?

david-arm replied:
I've kept it as part of the loop, because even the previous loop checking the size for the apple preferences bailed out on call instructions. It seems a bit inefficient to walk the loop twice, so I've renamed it to `canUnrollLoopWithinBudget`. I've also move the `findStringMetadataForLoop` out into `shouldUnrollMultiExitLoop`, since there's no need to check here.

fhahn replied:
Ok that's fine, although now we potentially check `shouldUnrollLoopWithInstruction` 3 times. Would it be better and more efficient to just check this once, before `// Apply subtarget-specific unrolling preferences`?

david-arm committed:
Address review comments

david-arm commented:
I plan to land this patch this morning if nobody has any objections?

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4542,43 +4631,18 @@ getAppleRuntimeUnrollPreferences(Loop *L, ScalarEvolution &SE,
   if (!L->isInnermost() || L->getNumBlocks() > 8)
     return;
 
+  // This is handled by common code.
```
```suggestion
  // Loops with multiple exits are handled by common code.
```

david-arm replied:
Done

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4528,6 +4528,95 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  if (MaxTC > 0 && MaxTC <= 32)
+    return false;
+
+  if (findStringMetadataForLoop(L, "llvm.loop.isvectorized"))
+    return false;
```
I think this is probably dead code, as the current version of `canUnrollLoopWithinBudget` forbids loops with vector instructions. This is a change to the current behavior for the Apple unrolling preferences, which allowed loops with vector code (e.g. intrinsics), but not loops vectorized by LV, which should already be unrolled by LV.

If there are enough vector execution units, I don't think there's a need to dis-allow vector instructions from unrolling, but that may require more target info, so keeping the bail-out is fine for now in `shouldUnrollLoopWithInstruction` is fine.

david-arm replied:
I've removed it anyway

fhahn replied:
LV should disable runtime unrolling explicitly via metadata and after running LV only runtime unrolling should run. I guess in theory there may be edge-cases where we fail to fully unroll initially and maybe fully unroll after vectorization, but that should probably be fine

david-arm committed:
Address review comments

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4687,6 +4746,16 @@ void AArch64TTIImpl::getUnrollingPreferences(Loop *L, ScalarEvolution &SE,
+  // unrolling.
+  for (auto *BB : L->getBlocks()) {
+    for (auto &I : *BB) {
+      if (!shouldUnrollLoopWithInstruction(I, *this))
```
nit: now that this has a single user, could be inlined again

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4528,6 +4528,88 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
   }
 }
 
+static bool shouldUnrollLoopWithInstruction(Instruction &I,
```
If not inlined back, would be good to doucment.

fhahn reviewed (approved):
LGTM, thanks

david-arm committed:
Address review comments
